### PR TITLE
RSDK-4810-plan-executor

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/RSDK-4810) 

- Delete motion/builtin/moveAttempt.go & fold it into motion/builtin/moveRequest.go. Move attempts were originally intended to be made multiple times per request. Given that now there is only ever a single attempt per request, they should simply be the same type.
- Add `Waypoints`, `PlanResp`, `PlanExecutorConstructor`, and `PlannerExecutor` types to motion/builtin/state
- Change `motion.builtin.moveRequest{}` to conform to state.PlannerExecutor interface
- Change `motion.MoveOnGlobe` & `motion.MoveOnMap` to call `state.PlannerExecutor` interface methods rather than old  `moveRequest` methods.
- Simplify `motion.builtin.moveRequest` constructor
- Change `motion.builtin.moveRequest` to distinguish between `MoveOnMap` and `MoveOnGlobe` requests (important for later PRs where we will need to be able to output `GeoPoses` for `MoveOnGlobe` but not for `MoveOnMap`
- Change moveRequest methods to propagate the replan reason any time the caller should replan

Manual testing:
1. `MoveOnGlobe`: Tested in gazebo using the nav service:
https://github.com/viamrobotics/rdk/assets/5927876/5010c77b-556e-488f-a1e8-d994527df982
2. `MoveOnMap`: Tested on a the SLAM bot

### CODE REVIEW

File: services/motion/builtin/move_request.go [hunk 1]
```diff
@@ -276,48 +327,60 @@ func newValidatedMotionCfg(motionCfg *motion.MotionConfiguration) (*validatedMot
...
+		return nil, err
+	}
+
+	if valExtra.maxReplans >= 0 {
+		if replanCount > valExtra.maxReplans {
+			return nil, fmt.Errorf("exceeded maximum number of replans: %d", valExtra.maxReplans)
+		}
+	}
+
+	motionCfg, err := newValidatedMotionCfg(req.MotionCfg)
```
Comment (nicksanford) [inline services/motion/builtin/move_request.go:347]: add tests for these cases

File: services/motion/builtin/move_request.go [hunk 2]
```diff
@@ -41,34 +44,82 @@ type validatedMotionConfiguration struct {
...
+		}, nil
+	case requestTypeMoveOnGlobe:
+		return state.PlanResp{
+			Waypoints:  waypoints,
+			Motionplan: plan,
+		}, nil
+	case requestTypeUnspecified:
+		fallthrough
+	default:
+		return state.PlanResp{}, fmt.Errorf("invalid moveRequest.requestType: %d", mr.requestType)
```
Comment (nicksanford) [inline services/motion/builtin/move_request.go:121]: add test for these cases

File: services/motion/builtin/builtin.go [hunk 3]
```diff
@@ -248,40 +248,28 @@ func (ms *builtIn) MoveOnMap(
 	ms.mu.RLock()
 	defer ms.mu.RUnlock()
 	operation.CancelOtherWithLabel(ctx, builtinOpLabel)
-	valExtra, err := newValidatedExtra(extra)
-	if err != nil {
-		return false, err
-	}
 	// make call to motionplan
-	mr, err := ms.newMoveOnMapRequest(ctx, componentName, destination, slamName, valExtra)
```
Comment (nicksanford) [inline services/motion/builtin/builtin.go:256]: add test for this

File: services/motion/builtin/move_request.go [hunk 4]
```diff
@@ -41,78 +44,131 @@ type validatedMotionConfiguration struct {
...
+	origin            spatialmath.GeoPose
+	logger            logging.Logger
 	config            *validatedMotionConfiguration
 	planRequest       *motionplan.PlanRequest
 	seedPlan          motionplan.Plan
 	kinematicBase     kinematicbase.KinematicBase
 	obstacleDetectors map[vision.Service][]resource.Name
 	replanCostFactor  float64
+
+	// from move attempt
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:68]: Confused why this is here and we need to store these things?  Wouldnt they be created by the plannerExecutor constructor?

**Review by raybjork — CHANGES_REQUESTED** [review]
Looks good! I dont think theres anything big that needs to change, mostly just have questions surrounding the new types were introducing.

File: services/motion/builtin/state/state.go [hunk 5]
```diff
@@ -6,8 +6,54 @@ import (
...
 
+// Waypoints represent the waypoints of the plan.
+type Waypoints [][]referenceframe.Input
+
+// PlanResp is the response from Plan.
+type PlanResp struct {
+	Waypoints        Waypoints
+	Motionplan       motionplan.Plan
+	GeoPoses         []spatialmath.GeoPose
+	PosesByComponent []motion.PlanStep
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/state/state.go:23]: Lets talk about this, I'm not aware of any way we can cut down on this but would love to be wrong about it.

[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/state/state.go:23]: @raybjork turns out I need to keep the `Motionplan` if we want to be able to use the previous plan to inform the new plan. If you don't want that functionality I can cut this type but if you want that functionality I gotta keep it. Keep in mind that currently that feature is broken on main.

File: services/motion/builtin/state/state.go [hunk 6]
```diff
@@ -6,8 +6,54 @@ import (
...
+}
+
+// ExecuteResp is the response from Execute.
+type ExecuteResp struct {
+	// If true, the Execute function didn't reach the goal & the caller should replan
+	Replan bool
+	// Set if Replan is true, describes why reaplanning was triggered
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/state/state.go]: done

File: services/motion/builtin/state/state.go [hunk 7]
```diff
@@ -6,8 +6,54 @@ import (
...
+// PlanResp is the response from Plan.
+type PlanResp struct {
+	Waypoints        Waypoints
+	Motionplan       motionplan.Plan
+	GeoPoses         []spatialmath.GeoPose
+	PosesByComponent []motion.PlanStep
+}
+
+// ExecuteResp is the response from Execute.
+type ExecuteResp struct {
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/state/state.go]: fixed

File: services/motion/builtin/state/state.go [hunk 8]
```diff
@@ -6,8 +6,54 @@ import (
...
+// if replanning has occurred
+// replanCount is the number of times replanning has occurred,
+// zero the first time planning occurs.
+// R is a genric type which is able to be used to create a PlannerExecutor.
+type PlanExecutorConstructor[R any] func(
+	ctx context.Context,
+	req R,
+	seedPlan motionplan.Plan,
+	replanCount int,
+) (PlannerExecutor, error)
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/state/state.go:46]: explained

File: services/motion/builtin/builtin.go [hunk 9]
```diff
@@ -350,74 +344,43 @@ func (ms *builtIn) MoveOnGlobe(
...
 
-		if valExtra.maxReplans >= 0 {
-			replanCount++
-			if replanCount > valExtra.maxReplans {
-				return false, fmt.Errorf("exceeded maximum number of replans: %d", valExtra.maxReplans)
-			}
-		}
+		// replan
+		ms.logger.Warnf("Replanning triggered: Reason: %s\n", resp.ReplanReason)
+		replanCount++
```
[the remark this answers is not in the document]
Comment (nicksanford) [reply services/motion/builtin/builtin.go:396]: b/c we need to give up when we reach the max replan count

File: services/motion/builtin/move_request.go (hunk 4 above)
Comment (nicksanford) [reply services/motion/builtin/move_request.go]: discussed offline. We need them for the background polling go routines that trigger replanning

File: services/motion/builtin/move_request.go [hunk 10]
```diff
@@ -41,78 +44,127 @@ type validatedMotionConfiguration struct {
...
+		return state.PlanResponse{
+			Waypoints:  waypoints,
+			Motionplan: plan,
+		}, nil
+	case requestTypeMoveOnGlobe:
+		return state.PlanResponse{
+			Waypoints:  waypoints,
+			Motionplan: plan,
+		}, nil
+	case requestTypeUnspecified:
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:113]: I don't think theres a reason to have `requestTypeUnspecified`.  Anything that is something we dont expect will fallthrough to the error case on its on its own.

File: services/motion/builtin/move_request.go [hunk 11]
```diff
@@ -41,78 +44,127 @@ type validatedMotionConfiguration struct {
...
 		select {
 		case <-ctx.Done():
-			return moveResponse{}
+			return state.ExecuteResponse{}, nil
 		default:
 			mr.planRequest.Logger.Info(waypoints[i])
 			if err := mr.kinematicBase.GoToInputs(ctx, waypoints[i]); err != nil {
 				// If there is an error on GoToInputs, stop the component if possible before returning the error
+				mr.logger.Debugf("calling kinematicBase.Stop due to %s\n", err)
 				if stopErr := mr.kinematicBase.Stop(ctx, nil); stopErr != nil {
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:133]: Are you happy about providing this context? I know there were some questions about which one we should be using.

**Review by raybjork — CHANGES_REQUESTED** [review]

**Review by raybjork — APPROVED** [review]

**Pull request merged**